examples/templates/aws-devcontainer/main.tf

```diff
@@ -144,24 +216,57 @@ locals {
     echo "Docker is already installed."
   fi
 
+  # Set up Docker credentials
+  mkdir -p "/home/${local.linux_user}/.docker"
+  if [ -n "${local.docker_config_json_base64}" ]; then
+     # Write the Docker config JSON to disk if it is provided.
+     printf "%s" "${local.docker_config_json_base64}" | base64 -d | tee "/home/${local.linux_user}/.docker/config.json"
+  else
+    # Assume that we're going to use the instance IAM role to pull from the cache repo if we need to.
+    # Set up the ecr credential helper.
+    apt-get update -y && apt-get install -y amazon-ecr-credential-helper
+    mkdir -p .docker
+    printf '{"credsStore": "ecr-login"}' | tee "/home/${local.linux_user}/.docker/config.json"
+  fi
+  chown -R ${local.linux_user}:${local.linux_user} "/home/${local.linux_user}/.docker"
+
+  # Write the container env to disk.
+  printf "%s" "${local.docker_env_list_base64}" | base64 -d | tee "/home/${local.linux_user}/env.txt"
+
   # Start envbuilder
-  docker run --rm \
+  sudo -u coder docker run \
+    --rm \
+    --net=host \
     -h ${lower(data.coder_workspace.me.name)} \
     -v /home/${local.linux_user}/envbuilder:/workspaces \
-    -e CODER_AGENT_TOKEN="${try(coder_agent.dev[0].token, "")}" \
-    -e CODER_AGENT_URL="${data.coder_workspace.me.access_url}" \
-    -e GIT_URL="${data.coder_parameter.repo_url.value}" \
-    -e INIT_SCRIPT="echo ${base64encode(try(coder_agent.dev[0].init_script, ""))} | base64 -d | sh" \
-    -e FALLBACK_IMAGE="codercom/enterprise-base:ubuntu" \
-    ghcr.io/coder/envbuilder
+    -v /var/run/docker.sock:/var/run/docker.sock \
+    --env-file /home/${local.linux_user}/env.txt \
+    ${local.builder_image}
   --//--
   EOT
 }
 
+# Check for the presence of a prebuilt image in the cache repo
+# that we can use instead.
+resource "envbuilder_cached_image" "cached" {
+  count         = var.cache_repo == "" ? 0 : data.coder_workspace.me.start_count
+  builder_image = local.devcontainer_builder_image
+  git_url       = data.coder_parameter.repo_url.value
+  cache_repo    = var.cache_repo
+  extra_env     = local.envbuilder_env
+}
+
+# This is useful for debugging the startup script. Left here for reference.
+# resource local_file "startup_script" {
+#   content  = local.user_data
+#   filename = "${path.module}/user_data.txt"
+# }
+
 resource "aws_instance" "vm" {
-  ami               = data.aws_ami.ubuntu.id
-  availability_zone = "${module.aws_region.value}a"
-  instance_type     = data.coder_parameter.instance_type.value
+  ami                  = data.aws_ami.ubuntu.id
+  availability_zone    = local.aws_availability_zone
+  instance_type        = data.coder_parameter.instance_type.value
+  iam_instance_profile = try(data.aws_iam_instance_profile.vm_instance_profile[0].name, null)
   root_block_device {
     volume_size = 30
```

Comment:
Should we also make the root disk size a parameter?

Reply:
I'd like to keep this a separate change.